### PR TITLE
added CRC integration draft

### PR DESCRIPTION
**Pull request by matbesancon**
PR showing the use of DiffOpt with ChainRules in a docs page

### CODE REVIEW

File: docs/make.jl [hunk 1]
```diff
@@ -19,13 +19,14 @@ makedocs(;
...
             "Solving conic with PSD and SOC constraints" => "solve-conic-1.md",
             "Differentiating a simple QP by hand" => "matrix-inversion-manual.md",
-            "Sensitivity Analysis SVM" => "sensitivity-analysis-svm.md"
-        ]
+            "Sensitivity Analysis SVM" => "sensitivity-analysis-svm.md",
+            "ChainRules integration" => "chainrules.md",
+        ],
     ],
     strict = true,  # See https://github.com/JuliaOpt/JuMP.jl/issues/1576
-    repo="https://github.com/jump-dev/DiffOpt.jl/blob/{commit}{path}#L{line}",
```
Comment (matbesancon) [inline docs/make.jl:26]: @be-apt I didn't get why we have this here? The commit/path/line part

**Comment by matbesancon** [comment]
also there are currently a lot of errors in the documentation:
```
┌ Warning: failed to run `@example` block in src/chainrules.md:3-7
│ ```@example 1
│ using DiffOpt, Plots, JuMP
│ import MathOptInterface, Clp
│ const MOI = MathOptInterface
│ ```
│   c.value =
│    ArgumentError: Package DiffOpt not found in current path:
│    - Run `import Pkg; Pkg.add("DiffOpt")` to install the DiffOpt package.
│    
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:9-11
│ ```@example 1
│ using ChainRulesCore
│ ```
│   c.value =
│    ArgumentError: Package ChainRulesCore not found in current path:
│    - Run `import Pkg; Pkg.add("ChainRulesCore")` to install the ChainRulesCore package.
│    
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:28-83
│ ```@example 1
│ function unit_commitment(load1_demand, load2_demand, gen_costs, noload_costs; model = Model(Clp.Optimizer), silent=false)
│     MOI.set(model, MOI.Silent(), silent)
│     
│     ## Problem data
│     unit_codes = [1, 2] # Generator identifiers
│     load_names = ["Load1", "Load2"] # Load identifiers
│     n_periods = 4 # Number of time periods
│     Pmin = Dict(1 => fill(0.5, n_periods), 2 => fill(0.5, n_periods)) # Minimum power output (pu)
│     Pmax = Dict(1 => fill(3.0, n_periods), 2 => fill(3.0, n_periods)) # Maximum power output (pu)
│     RR = Dict(1 => 0.25, 2 => 0.25) # Ramp rates (pu/min)
│     P0 = Dict(1 => 0.0, 2 => 0.0) # Initial power output (pu)
│     D = Dict("Load1" => load1_demand, "Load2" => load2_demand) # Demand (pu)
│     Cp = Dict(1 => gen_costs[1], 2 => gen_costs[2]) # Generation cost coefficient ($/pu)
│     Cnl = Dict(1 => noload_costs[1], 2 => noload_costs[2]) # No-load cost ($)
│ 
│     ## Variables
│     # Note: u represents the activation of generation units.
│     # Would be binary in the typical UC problem, relaxed here to u ∈ [0,1]
│     # for a linear relaxation.
│     @variable(model, 0 <= u[g in unit_codes, t in 1:n_periods] <= 1) # Commitment
│     @variable(model, p[g in unit_codes, t in 1:n_periods] >= 0) # Power output (pu)
│     
│     ## Constraints
│     
│     # Energy balance
│     @constraint(
│         model,
│         energy_balance_cons[t in 1:n_periods],
│         sum(p[g, t] for g in unit_codes) == sum(D[l][t] for l in load_names),
│     )
│     
│     # Generation limits
│     @constraint(model, [g in unit_codes, t in 1:n_periods], Pmin[g][t] * u[g, t] <= p[g, t])
│     @constraint(model, [g in unit_codes, t in 1:n_periods], p[g, t] <= Pmax[g][t] * u[g, t])
│     
│     # Ramp rates
│     @constraint(model, [g in unit_codes, t in 2:n_periods], p[g, t] - p[g, t - 1] <= 60 * RR[g])
│     @constraint(model, [g in unit_codes], p[g, 1] - P0[g] <= 60 * RR[g])
│     @constraint(model, [g in unit_codes, t in 2:n_periods], p[g, t - 1] - p[g, t] <= 60 * RR[g])
│     @constraint(model, [g in unit_codes], P0[g] - p[g, 1] <= 60 * RR[g])
│     
│     # Objective
│     @objective(
│         model,
│         Min,
│         sum((Cp[g] * p[g, t]) + (Cnl[g] * u[g, t]) for g in unit_codes, t in 1:n_periods),
│     )
│     
│     optimize!(model)
│     # asserting finite optimal value
│     @assert termination_status(model) == MOI.OPTIMAL
│     # converting to dense matrix
│     return JuMP.value.(p.data)
│ end
│ ```
│   c.value =
│    LoadError: UndefVarError: @variable not defined
│    in expression starting at none:20
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:85-93
│ ```@example 1
│ m = Model(Clp.Optimizer)
│ @show unit_commitment(
│     [1.0, 1.2, 1.4, 1.6], [1.0, 1.2, 1.4, 1.6],
│     [1000.0, 1500.0], [500.0, 1000.0],
│     model=m, silent=true
│ )
│ println(m)
│ ```
│   c.value = UndefVarError: Clp not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:95-99
│ ```@example 1
│ demand_values = 0.05:0.05:3.0
│ pvalues = [unit_commitment([1.0, di, 1.4, 1.6], [1.0, di, 1.4, 1.6], [1000.0, 1500.0], [500.0, 1000.0], silent=true) for di in demand_values]
│ pflat = [getindex.(pvalues, i) for i in eachindex(pvalues[1])]
│ ```
│   c.value = UndefVarError: unit_commitment not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:101-104
│ ```@example 1
│ scatter(demand_values, pflat, size=(1500,1000), legendfontsize=20, markersize=20)
│ title!("Generation at different time frames and generators for a single variation")
│ ```
│   c.value = UndefVarError: scatter not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:116-153
│ ```@example 1
│ function ChainRulesCore.frule(
│         (_, Δload1_demand, Δload2_demand, Δgen_costs, Δnoload_costs),
│         ::typeof(unit_commitment),
│         load1_demand, load2_demand, gen_costs, noload_costs;
│         optimizer=Clp.Optimizer,
│         )
│     # creating the UC model with a DiffOpt optimizer wrapper around Clp
│     model = Model(() -> diff_optimizer(optimizer))
│     # building and solving the main model
│     pv = unit_commitment(load1_demand, load2_demand, gen_costs, noload_costs, model=model)
│     energy_balance_cons = model[:energy_balance_cons]
│ 
│     # Setting some perturbation of the right-hand side of the energy balance constraints
│     # the RHS is equal to the sum of load demands at each period.
│     # the corresponding perturbation are set accordingly as the set of perturbations of the two loads
│     MOI.set.(
│         model,
│         DiffOpt.ForwardIn{DiffOpt.ConstraintConstant}(), energy_balance_cons,
│         [d1 + d2 for (d1, d2) in zip(Δload1_demand, Δload2_demand)],
│     )
│ 
│     p = model[:p]
│     u = model[:u]
│ 
│     # setting the perturbation of the linear objective
│     for t in size(p, 2)
│         MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), p[1,t], Δgen_costs[1])
│         MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), p[2,t], Δgen_costs[2])
│         MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), u[1,t], Δnoload_costs[1])
│         MOI.set(model, DiffOpt.ForwardIn{DiffOpt.LinearObjective}(), u[2,t], Δnoload_costs[2])
│     end
│     DiffOpt.forward(JuMP.backend(model))
│     # querying the corresponding perturbation of the decision
│     Δp = MOI.get.(model, DiffOpt.ForwardOut{MOI.VariablePrimal}(), p)
│     return (pv, Δp.data)
│ end
│ ```
│   c.value = UndefVarError: ChainRulesCore not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:155-171
│ ```@example 1
│ load1_demand = [1.0, 1.0, 1.4, 1.6]
│ load2_demand = [1.0, 1.0, 1.4, 1.6]
│ gen_costs = [1000.0, 1500.0]
│ noload_costs = [500.0, 1000.0]
│ 
│ Δload1_demand = 0 * load1_demand
│ Δload1_demand[2] = 1.0
│ Δload2_demand = 0 * load2_demand
│ Δgen_costs = 0 * gen_costs
│ Δnoload_costs = 0 * noload_costs
│ (pv, Δpv) = ChainRulesCore.frule(
│     (nothing, Δload1_demand, Δload2_demand, Δgen_costs, Δnoload_costs),
│     unit_commitment,
│     load1_demand, load2_demand, gen_costs, noload_costs
│ )
│ ```
│   c.value = UndefVarError: ChainRulesCore not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:166-168
│ ```@example 1
│ Δpv
│ ```
│   c.value = UndefVarError: Δpv not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:183-216
│ ```@example 1
│ function ChainRulesCore.rrule(
│         ::typeof(unit_commitment),
│         load1_demand, load2_demand, gen_costs, noload_costs;
│         optimizer=Clp.Optimizer,
│         silent=false)
│     model = Model(() -> diff_optimizer(optimizer))
│     # solve the forward UC problem
│     pv = unit_commitment(load1_demand, load2_demand, gen_costs, noload_costs, model=model, silent=silent)
│     function pullback_unit_commitment(pb)
│         p = model[:p]
│         u = model[:u]
│         energy_balance_cons = model[:energy_balance_cons]
│ 
│         MOI.set.(model, DiffOpt.BackwardIn{MOI.VariablePrimal}(), p, pb)
│         DiffOpt.backward(JuMP.backend(model))
│ 
│         # computing derivative wrt linear objective costs
│         dgen_costs = similar(gen_costs)
│         dgen_costs[1] = sum(MOI.get.(model, DiffOpt.BackwardOut{DiffOpt.LinearObjective}(), p[1,:]))
│         dgen_costs[2] = sum(MOI.get.(model, DiffOpt.BackwardOut{DiffOpt.LinearObjective}(), p[2,:]))
│ 
│         dnoload_costs = similar(noload_costs)
│         dnoload_costs[1] = sum(MOI.get.(model, DiffOpt.BackwardOut{DiffOpt.LinearObjective}(), u[1,:]))
│         dnoload_costs[2] = sum(MOI.get.(model, DiffOpt.BackwardOut{DiffOpt.LinearObjective}(), u[2,:]))
│         
│         # computing derivative wrt constraint constant
│         dload1_demand = MOI.get.(model, DiffOpt.BackwardOut{DiffOpt.ConstraintConstant}(), energy_balance_cons)
│         dload2_demand = copy(dload1_demand)
│         return (dload1_demand, dload2_demand, dgen_costs, dnoload_costs)
│     end
│     return (pv, pullback_unit_commitment)
│ end
│ ```
│   c.value = UndefVarError: ChainRulesCore not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/chainrules.md:218-231
│ ```@example 1
│ (pv, pullback_unit_commitment) = ChainRulesCore.rrule(
│     unit_commitment,
│     load1_demand, load2_demand, gen_costs, noload_costs,
│     model = Model(() -> diff_optimizer(Clp.Optimizer)),
│     silent=true,
│ )
│ dpv = 0 * pv
│ dpv[1,2] = 1
│ dargs = pullback_unit_commitment(dpv)
│ (dload1_demand, dload2_demand, dgen_costs, dnoload_costs) = dargs
│ @show dload1_demand
│ @show dload2_demand;
│ ```
│   c.value = UndefVarError: Model not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:24-34
│ ```@example 1
│ import Random
│ using Test
│ import SCS
│ import Plots
│ using DiffOpt
│ using LinearAlgebra
│ import MathOptInterface
│ const MOI = MathOptInterface
│ nothing # hide
│ ```
│   c.value =
│    ArgumentError: Package DiffOpt not found in current path:
│    - Run `import Pkg; Pkg.add("DiffOpt")` to install the DiffOpt package.
│    
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:47-56
│ ```@example 1
│ model = diff_optimizer(SCS.Optimizer) 
│ MOI.set(model, MOI.Silent(), true)
│ 
│ # add variables
│ l = MOI.add_variables(model, N)
│ w = MOI.add_variables(model, D)
│ b = MOI.add_variable(model)
│ nothing # hide
│ ```
│   c.value = UndefVarError: diff_optimizer not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:59-85
│ ```@example 1
│ MOI.add_constraint(
│     model,
│     MOI.VectorAffineFunction(
│         MOI.VectorAffineTerm.(1:N, MOI.ScalarAffineTerm.(1.0, l)), zeros(N),
│     ), 
│     MOI.Nonnegatives(N),
│ )
│ 
│ # define the whole matrix Ax, it'll be easier then
│ # refer https://discourse.julialang.org/t/solve-minimization-problem-where-constraint-is-the-system-of-linear-inequation-with-mathoptinterface-efficiently/23571/4
│ Ax = Matrix{MOI.ScalarAffineTerm{Float64}}(undef, N, D+2)
│ for i in 1:N
│     Ax[i, :] = MOI.ScalarAffineTerm.([1.0; y[i]*X[i,:]; y[i]], [l[i]; w; b])
│ end
│ terms = MOI.VectorAffineTerm.(1:N, Ax)
│ f = MOI.VectorAffineFunction(
│     vec(terms),
│     -ones(N),
│ )
│ cons = MOI.add_constraint(
│     model,
│     f,
│     MOI.Nonnegatives(N),
│ )
│ nothing # hide
│ ```
│   c.value = UndefVarError: MOI not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:88-102
│ ```@example 1
│ objective_function = MOI.ScalarAffineFunction(
│                         MOI.ScalarAffineTerm.(ones(N), l),
│                         0.0,
│                     )
│ MOI.set(model, MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}(), objective_function)
│ MOI.set(model, MOI.ObjectiveSense(), MOI.MIN_SENSE)
│ 
│ MOI.optimize!(model)
│ 
│ loss = MOI.get(model, MOI.ObjectiveValue())
│ wv = MOI.get(model, MOI.VariablePrimal(), w)
│ bv = MOI.get(model, MOI.VariablePrimal(), b)
│ nothing # hide
│ ```
│   c.value = UndefVarError: MOI not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:106-116
│ ```@example 1
│ # build SVM points
│ svm_x = [0.0, 3.0]
│ svm_y = (-bv .- wv[1] * svm_x )/wv[2]
│ 
│ p = Plots.scatter(X[:,1], X[:,2], color = [yi > 0 ? :red : :blue for yi in y], label = "")
│ Plots.yaxis!(p, (-2, 4.5))
│ Plots.plot!(p, svm_x, svm_y, label = "loss = $(round(loss, digits=2))", width=3)
│ Plots.savefig("svm_separating.svg")
│ nothing # hide
│ ```
│   c.value = UndefVarError: bv not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:165-199
│ ```@example 1
│ ∇ = Float64[]
│ dy = zeros(N)
│ 
│ # begin differentiating
│ for Xi in 1:N
│     dy[Xi] = 1.0  # set
│     
│     MOI.set(
│         model,
│         DiffOpt.ForwardIn{DiffOpt.ConstraintCoefficient}(), 
│         b, 
│         cons, 
│         dy
│     )
│     
│     DiffOpt.forward(model)
│     
│     dw = MOI.get.(
│         model,
│         DiffOpt.ForwardOut{MOI.VariablePrimal}(), 
│         w
│     ) 
│     db = MOI.get(
│         model,
│         DiffOpt.ForwardOut{MOI.VariablePrimal}(), 
│         b
│     ) 
│     push!(∇, norm(dw) + norm(db))
│     
│     dy[Xi] = 0.0  # reset the change made above
│ end
│ LinearAlgebra.normalize!(∇)
│ nothing # hide
│ ```
│   c.value = UndefVarError: MOI not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:202-212
│ ```@example 1
│ p2 = Plots.scatter(
│     X[:,1], X[:,2], 
│     color = [yi > 0 ? :red : :blue for yi in y], label = "",
│     markersize = ∇ * 20,
│ )
│ Plots.yaxis!(p2, (-2, 4.5))
│ Plots.plot!(p2, svm_x, svm_y, label = "loss = $(round(loss, digits=2))", width=3)
│ Plots.savefig("sensitivity2.svg")
│ nothing # hide
│ ```
│   c.value = UndefVarError: loss not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
┌ Warning: failed to run `@example` block in src/sensitivity-analysis-svm.md:258-268
│ ```@example 1
│ p3 = Plots.scatter(
│     X[:,1], X[:,2], 
│     color = [yi > 0 ? :red : :blue for yi in y], label = "",
│     markersize = ∇ * 20,
│ )
│ Plots.yaxis!(p3, (-2, 4.5))
│ Plots.plot!(p3, svm_x, svm_y, label = "loss = $(round(loss, digits=2))", width=3)
│ Plots.savefig(p3, "sensitivity3.svg")
│ nothing # hide
│ ```
│   c.value = UndefVarError: loss not defined
└ @ Documenter.Expanders ~/.julia/packages/Documenter/bFHi4/src/Expanders.jl:563
```

**Comment by matbesancon** [comment]
@blegat your last PR fixed the issue I had with the docs. However I get an UnsupportedAttribute exception:

```julia
┌ Warning: failed to run `@example` block in src/chainrules.md:167-183
│ ```@example 1
│ load1_demand = [1.0, 1.0, 1.4, 1.6]
│ load2_demand = [1.0, 1.0, 1.4, 1.6]
│ gen_costs = [1000.0, 1500.0]
│ noload_costs = [500.0, 1000.0]
│ 
│ Δload1_demand = 0 * load1_demand
│ Δload1_demand[2] = 1.0
│ Δload2_demand = 0 * load2_demand
│ Δgen_costs = 0 * gen_costs
│ Δnoload_costs = 0 * noload_costs
│ (pv, Δpv) = ChainRulesCore.frule(
│     (nothing, Δload1_demand, Δload2_demand, Δgen_costs, Δnoload_costs),
│     unit_commitment,
│     load1_demand, load2_demand, gen_costs, noload_costs
│ )
│ ```
│   c.value = MathOptInterface.UnsupportedAttribute{DiffOpt.ForwardInConstraint}: Attribute DiffOpt.ForwardInConstraint() is not supported by the model.
```

File: docs/src/chainrules.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+In this example, we will demonstrate the integration of DiffOpt with
+[ChainRulesCore.jl](https://juliadiff.org/ChainRulesCore.jl/stable/),
+the library allowing the definition of derivatives for functions
+that can then be used by automatic differentiation systems.
+
+```@example 1
+using DiffOpt, Plots, JuMP
+import MathOptInterface, Clp
+const MOI = MathOptInterface
```
Comment (blegat) [inline docs/src/chainrules.md:11]: This is already exported by JuMP, these last 2 lines can be removed

File: docs/src/chainrules.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        DiffOpt.ForwardInConstraint(), energy_balance_cons,
+        [d1 + d2 for (d1, d2) in zip(Δload1_demand, Δload2_demand)],
+    )
+
+    p = model[:p]
+    u = model[:u]
+
+    # setting the perturbation of the linear objective
+    for t in size(p, 2)
+        MOI.set(model, DiffOpt.ForwardInObjective(), p[1,t], Δgen_costs[1])
```
Comment (blegat) [inline docs/src/chainrules.md:155]: These should be updated, ForwardInConstraint and ForwardInObjective now takes the function as argument.
See
https://github.com/jump-dev/DiffOpt.jl/blob/2a373d07b1c45b02b5d03c9699ee6b23f70297bc/examples/chainrules.jl#L93
https://github.com/jump-dev/DiffOpt.jl/blob/2a373d07b1c45b02b5d03c9699ee6b23f70297bc/examples/chainrules.jl#L100-L101

File: docs/src/chainrules.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
+# ChainRules integration
```
Comment (joaquimg) [inline docs/src/chainrules.md:1]: shouldnt we do this with a tiny toy example with 2 variables and 2 constraints max?
The whole concept of UC might push people away.
Then we do a UC version in a tutorial.

Comment (matbesancon) [reply docs/src/chainrules.md:1]: Yes I agree, I think we can add another one that's simpler? This one is still interesting because non-trivial

Comment (joaquimg) [reply docs/src/chainrules.md:1]: sure we can have both!
Once we have the simple one, I would put the simple in the manual and this complex one in the tutorials.

Comment (matbesancon) [reply docs/src/chainrules.md]: tutorials still in these docs right?

**Comment by blegat** [comment]
> However I get an UnsupportedAttribute exception:

Do you still get it ? I see the CI is green

**Review by blegat — APPROVED** [review]

**Comment by matbesancon** [comment]
no this is all fine now

**Pull request merged**